inventory/byo/hosts.origin.example

```diff
@@ -89,6 +89,21 @@ openshift_release=v1.4
 # Skip upgrading Docker during an OpenShift upgrade, leaves the current Docker version alone.
 # docker_upgrade=False
 
+
+# Upgrade Hooks
+#
+# Hooks are available to run custom tasks at various points during a cluster
+# upgrade. Each hook should point to a file with Ansible tasks defined. Suggest using
+# absolute paths, if not the path will be treated as relative to the file where the
+# hook is actually used.
+#
+# Tasks to run before each master is upgraded:
+# openshift_upgrade_pre_master_hook=/usr/share/custom/pre_master.yml
```

Comment:
Hello and thank you for quick response!
We just finished successfully upgrade errata for latest 3.2->3.2 version. I want share real world scenario we faced.

We have 12 byo servers running HA Openshift 3.2. All 3 masters also host pods. Each one of 12 nodes has GlusterFS pod mounting local disks on server. When we run upgrade on staging environment each master upgrade also restarted node service, so after all 3 masters were upgraded, we lost whole gluster cluster since auto-heal was not so fast as upgrade moved on (we have replication factor of 3 in gluster). Additionally, upgrade decided to leave master nodes unscheduled even before we started all of them were scheduled and running pods. Also, upgrade detected HA and serialized master upgrade - in our case its problem too. So what we did before upgrade:

```
- hosts: primary_master
  tasks:
  - name: Alter stock update script1
    blockinfile:
      dest: /usr/share/ansible/openshift-ansible/playbooks/common/openshift-cluster/upgrades/upgrade.yml
      insertafter: '\s\swhen\:\ (|\"\{\{\s)inventory_hostname\ in\ groups\.oo_nodes_to_config\ and\ (openshift\.node\.|was_)schedulable\ \|\ bool(|\s\}\}\")$'
      marker: "#{mark}#INSERTED MANAGED BLOCK###"
      content: "  - command: >\n      {{ '{{' }} openshift.common.admin_binary {{ '}}' }} manage-node {{ '{{' }} openshift.common.hostname | lower {{ '}}' }} --schedulable=true\n    delegate_to: \"{{ '{{' }} groups.oo_first_master.0 {{ '}}' }}\"\n  - include: {{ ansible_scripts_dir }}/ansible-openshift/vars/custom_block.yml"
      state: present
    register: insert_result1

  - assert:
      that:
        - 'insert_result1.changed'
        - 'insert_result1.msg == "Block inserted"'
      msg: "Will not continue since we unable to alter original upgrade script1"

  - name: Alter stock update script2
    blockinfile:
      dest: /usr/share/ansible/openshift-ansible/playbooks/common/openshift-cluster/upgrades/rpm_upgrade.yml
      marker: "#{mark}#INSERTED MANAGED BLOCK###"
      content: "- command: >\n      {{ '{{' }} openshift.common.admin_binary {{ '}}' }} manage-node {{ '{{' }} openshift.common.hostname | lower {{ '}}' }} --schedulable=true\n    delegate_to: \"{{ '{{' }} groups.oo_first_master.0 {{ '}}' }}\"\n- include: {{ ansible_scripts_dir }}/ansible-openshift/vars/custom_block.yml"
      state: present
    register: insert_result2

  - assert:
      that:
        - 'insert_result2.changed'
        - 'insert_result2.msg == "Block inserted"'
      msg: "Will not continue since we unable to alter original upgrade script2"

  - name: Alter stock update script3
    blockinfile:
      dest: /usr/share/ansible/openshift-ansible/playbooks/common/openshift-cluster/upgrades/upgrade.yml
      marker: "#{mark}#INSERTED MANAGED BLOCK2###"
      content: "  serial: 1"
      insertafter: '\s\s\s\swhen\:\ (|\"\{\{\s)ca_crt_stat\.stat\.isreg\ and\ not ca_bundle_stat\.stat\.exists(|\s\}\}\")$'
      state: present
    register: insert_result3

  - assert:
      that:
        - 'insert_result3.changed'
        - 'insert_result3.msg == "Block inserted"'
      msg: "Will not continue since we unable to alter original upgrade script3"
```
custom_block.yml has tasks we run after each gluster pod is restarted to make sure cluster heal completed on all pods. We put long retry and this helped us because we had major OS problem after one of nodes was upgraded. We even restarted this server several times and reinstalled few packages to solve the problem. Eventually when pod on this server was up, the retry succeeded and whole upgrade finished as expected.
 Just in any case here are contents of custom_block.yml:

```
- name: Delay to make sure nodes status updated
  pause: minutes=1

- set_fact: project="storage-project"

# This will check all nodes are ready
- name: Check Openshift nodes
  shell: oc get nodes | grep NotReady | wc -l
  register: ose_nodes_up
  retries: 43200
  delay: 60
  failed_when: "ose_nodes_up.stdout != '0'"
  until: ose_nodes_up.stdout == "0"
  delegate_to: "{{ groups.oo_first_master.0 }}"

# This will check if one node is SchedulingDisabled
- name: Check Openshift node SchedulingDisabled
  shell: oc get nodes | grep SchedulingDisabled | wc -l
  register: ose_scheduling_disabled
  retries: 43200
  delay: 60
  failed_when: "ose_scheduling_disabled.rc != 0"
  until: ose_scheduling_disabled.rc == 0
  delegate_to: "{{ groups.oo_first_master.0 }}"

# This will check all Gluster pods are ready
- name: Check Gluster pods
  shell: oc get pods --namespace={{ project }} | grep "0/1" | wc -l
  register: ose_pods_up
  retries: 43200
  delay: 60
  failed_when: "ose_pods_up.stdout != '0'"
  until: ose_pods_up.stdout == "0"
  delegate_to: "{{ groups.oo_first_master.0 }}"
  when: "ose_scheduling_disabled.stdout == '0'"

- name: Get GlusterFS pods
  shell: oc get pods --namespace={{ project }} -o name | grep glusterfs | grep -v Completed | grep -v Terminating | cut -d'/' -f2
  register: pods
  delegate_to: "{{ groups.oo_first_master.0 }}"
  when: "ose_scheduling_disabled.stdout == '0'"

- name: Build dictionary of pods
  set_fact:
    pods_dict: "{{ pods_dict|default({}) | combine( {item.split('.').0: item} ) }}"
  with_items: "{{ pods.stdout_lines | default([]) }}"
  when: "ose_scheduling_disabled.stdout == '0'"

- name: Get master-pod-volume pairs to check
  shell: oc exec {{ pods_dict[item] }} gluster volume list --namespace={{ project }}
  register: volumes
  with_items:
    - "{{ pods_dict.keys() | default([]) }}"
  delegate_to: "{{ groups.oo_first_master.0 }}"
  retries: 43200
  delay: 60
  until: volumes.rc == 0
  when: "ose_scheduling_disabled.stdout == '0'"

# This will check that all nodes connected to Gluster
- name: Check Gluster peers
  shell: oc exec {{ pods_dict[item] }} gluster peer status --namespace={{ project }} | grep "(Disconnected)" | wc -l
  with_items:
    - '{{ pods_dict.keys() | default([]) }}'
  register: gluster_peers
  failed_when: gluster_peers.stdout != '0'
  retries: 43200
  delay: 60
  until: gluster_peers.stdout == '0'
  when: "ose_scheduling_disabled.stdout == '0'"
  delegate_to: "{{ groups.oo_first_master.0 }}"

- name: Build dictionary of volumes
  set_fact:
    vols_dict: "{{ vols_dict|default({}) | combine( {item.item: { 'pod': item.item, 'volumes': item.stdout_lines} } ) }}"
  with_items: "{{ volumes.results | default([]) }}"
  when: 'ose_scheduling_disabled.stdout == "0" and item.stdout != ""'
  delegate_to: "{{ groups.oo_first_master.0 }}"

# This will check if there is any ongoing task on any of volumes on any of Gluster pods. If it finds task it will loop for a month until task completes
- name: Check Gluster tasks
  shell: oc exec {{ pods_dict[item.0.pod] }} gluster volume status {{ item.1 }} tasks --namespace={{ project }} | grep "There are no active volume tasks" | wc -l
  with_subelements:
    - '{{ vols_dict | default([]) }}'
    - volumes
  register: gluster_tasks
  failed_when: gluster_tasks.stdout != "1"
  retries: 43200
  delay: 60
  until: gluster_tasks.stdout == "1"
  when: "ose_scheduling_disabled.stdout == '0'"
  delegate_to: "{{ groups.oo_first_master.0 }}"

# This will check if any Gluster bricks are healing
- name: Check bricks healing
  shell: oc exec {{ pods_dict[item.0.pod] }} gluster volume heal {{ item.1 }} info --namespace={{ project }} | grep "Number of entries" | grep -v " 0" | wc -l
  with_subelements:
    - '{{ vols_dict | default([]) }}'
    - volumes
  register: healing_tasks
  failed_when: healing_tasks.stdout != "0"
  retries: 43200
  delay: 60
  until: healing_tasks.stdout == "0"
  when: "ose_scheduling_disabled.stdout == '0'"
  delegate_to: "{{ groups.oo_first_master.0 }}"
```